flang/lib/Lower/IO.cpp

```diff
@@ -1401,16 +1410,94 @@ mlir::Value Fortran::lower::genReadStatement(
+      beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireUnit)>(loc, builder);
+      mlir::FunctionType beginFuncTy = beginFunc.getType();
+
+      auto unit = converter.genExprValue(
+          getExprInList<Fortran::parser::FileUnitNumber>(*list), loc);
+      auto un = builder.createConvert(loc, beginFuncTy.getInput(0), unit);
+      auto file = getDefaultFilename(builder, loc, beginFuncTy.getInput(1));
+      auto line = getDefaultLineNo(builder, loc, beginFuncTy.getInput(2));
+      // Append BeginXyz call arguments.  File name and line number are always
+      // last. For a 'BeginInquireUnit' the IO runtime API looks like below
+      // BeginXyz (ExternalUnit, const char *sourceFile = nullptr, int
+      // sourceLine = 0); accordingly match the arguments and types
+      llvm::SmallVector<mlir::Value, 4> beginArgs{un, file, line};
+
+      // Arguments are done; call the BeginXyz function.
+      cookie =
+          builder.create<mlir::CallOp>(loc, beginFunc, beginArgs).getResult(0);
+      // Handle remaining arguments in specifier list.
+      genConditionHandlerCall(converter, loc, cookie, *list, csi);
+
+    } else if (hasX<Fortran::parser::FileNameExpr>(*list)) {
+      // Determine which BeginXyz call to make.
+      beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireFile)>(loc, builder);
+      mlir::FunctionType beginFuncTy = beginFunc.getType();
+
+      auto file = converter.genExprValue(
+          getExprInList<Fortran::parser::FileNameExpr>(*list), loc);
+      // Helper to query [BUFFER, LEN].
+      Fortran::lower::CharacterExprHelper helper{builder, loc};
+      auto dataLen = helper.materializeCharacter(file);
+      auto buff =
+          builder.createConvert(loc, beginFuncTy.getInput(0), dataLen.first);
+      auto len =
+          builder.createConvert(loc, beginFuncTy.getInput(1), dataLen.second);
+      auto kindInt = helper.getCharacterKind(file.getType());
+      mlir::Value kindValue{
+          builder.createIntegerConstant(loc, beginFuncTy.getInput(2), kindInt)};
+      auto sourceFile =
+          getDefaultFilename(builder, loc, beginFuncTy.getInput(3));
+      auto line = getDefaultLineNo(builder, loc, beginFuncTy.getInput(4));
+
+      // Append BeginXyz call arguments.  File name and line number are always
+      // last. For a 'BeginFileNameExpr' the IO runtime API is 'BeginXyz (const
+      // char *, std::size_t, int kind,
+      //    const char *sourceFile = nullptr, int sourceLine = 0)'
+      // accordingly pass the arguments.
+      llvm::SmallVector<mlir::Value, 5> beginArgs = {
+          buff, len, kindValue, sourceFile, line,
+      };
+
+      // Arguments are done; call the BeginXyz function.
+      cookie =
+          builder.create<mlir::CallOp>(loc, beginFunc, beginArgs).getResult(0);
+      // Handle remaining arguments in specifier list.
+      genConditionHandlerCall(converter, loc, cookie, *list, csi);
+    }
+  } else if (const auto *ioLength{
```

Comment:
There are 3 variations of BeginInquireXxx that we need to support. Semantics should catch an INQUIRE that is ill-formed. Here, in lowering, we just need a simple structure of `if (case1) { ... } else if (case2) { ... } else { assert(case3); ... }`. The control-flow here has two fall-through paths that will just not emit a BeginInquireXxx at all if garbage input comes into lowering.

We don't want that because it will lead to incorrect code being silently generated by the compiler.